### PR TITLE
Fix addons.smearing RHF electron number

### LINKED ISSUE #2289: Smearing electron number bug
Smeared RHF with odd nelec gives incorrect electron number.

```
from pyscf import gto, scf
from pyscf.scf import addons

mol = gto.Mole()
mol.atom = '''
    7      0.   0  -0.7
    7      0.   0   0.7'''
mol.basis = 'cc-pvdz'
mol.charge = -1
mol.spin = 1
mol.build()
mf = scf.RHF(mol)
mf = addons.smearing(mf, sigma=0.1)
mf.kernel()
print(mf.mol.nelectron)
print(mf.mo_occ.sum())
```

```
converged SCF energy = -42.2391465363651
15
16.0
```

### PR DESCRIPTION
**Pull request by chillenb**
Fixes #2289.

### CODE REVIEW

**Comment by chillenb** [comment]
Just rebased off newest master branch.

**Comment by chillenb** [comment]
@sunqm it passes all the tests now thanks to the rebase -- do you have any thoughts?

File: pyscf/scf/addons.py [hunk 1]
```diff
@@ -79,11 +80,15 @@ def nelec_cost_fn(m, mo_es, nocc):
...
+    if not res.success:
+        logger.error(rec, f'addons._smearing_optimize failed: {res.message}')
+    if numpy.abs(mo_occs.sum() - nocc) > ntol:
+        logger.error(rec, f'addons._smearing_optimize failed: {mo_occs.sum()} != {nocc}')
     return mu, mo_occs
 
 def _get_fermi(mo_energy, nocc):
     mo_e_sorted = numpy.sort(mo_energy)
-    return mo_e_sorted[nocc-1]
+    return mo_e_sorted[math.ceil(nocc)-1]
```
Comment (fishjojo) [inline pyscf/scf/addons.py:91]: can be `return mo_e_sorted[numpy.ceil(nocc).astype(int)-1]`

File: pyscf/scf/addons.py [hunk 2]
```diff
@@ -79,11 +80,15 @@ def nelec_cost_fn(m, mo_es, nocc):
         options={'xtol': 1e-5, 'ftol': 1e-5, 'maxiter': 10000})
     mu = res.x
     mo_occs = f_occ(mu, mo_es, sigma)
+    if not res.success:
+        logger.error(rec, f'addons._smearing_optimize failed: {res.message}')
+    if numpy.abs(mo_occs.sum() - nocc) > ntol:
+        logger.error(rec, f'addons._smearing_optimize failed: {mo_occs.sum()} != {nocc}')
```
Comment (fishjojo) [inline pyscf/scf/addons.py:86]: These are not necessary in my opinion.

**Comment by chillenb** [comment]
@fishjojo 
Thanks for your suggestions! I've made the changes you requested.

**Review by fishjojo — APPROVED** [review]

File: pyscf/scf/addons.py
Comment (chillenb) [inline pyscf/scf/addons.py:1]: Thanks for cleaning that up for me. I appreciate it

**Pull request merged**